io/io/test/TBufferMerger.cxx

```diff
@@ -229,3 +229,63 @@ TEST(TBufferMerger, CheckTreeFillResults)
    RemoveFile("tbuffermerger_sequential.root");
    RemoveFile("tbuffermerger_parallel.root");
 }
+
+/**
+ * \test TBufferMerger, SetMaxTreeSize
+ * \brief Test to avoid issue #6523.
+ * 
+ * `TTree`'s default behaviour of changing the file it is attached to when reaching
+ * a size greater than `fgMaxTreeSize` doesn't fit in the design of TBufferMerger.
+ * The `TTree::Fill` method has been modified accordingly, avoiding this behaviour
+ * when the tree is attached to a TMemFile (thus also a TBufferMergerFile). This
+ * test tries to trigger the behaviour forcedly by calling `TTree::SetMaxTreeSize`
+ * but the TBufferMergerFile is never detached from the tree.
+ */
+TEST(TBufferMerger, SetMaxTreeSize)
+{
+   ROOT::EnableThreadSafety();
+
+   {
+      ROOT::Experimental::TBufferMerger merger{"tbuffermerger_setmaxtreesize.root"};
+
+      auto tbmfile = merger.GetFile(); // std::shared_ptr<TBufferMergerFile>
+
+      int nentries{20000};
+      int maxtreesize{1000};
+
+      TTree tree{"T", "SetMaxTreeSize(1000)"};
+      tree.SetMaxTreeSize(maxtreesize);
+      // The resetting of the kCleanup bit below is necessary to avoid leaving
+      // the management of this object to ROOT, which leads to a race condition
+      // that may cause a crash once all threads are finished and the final
+      // merge is happening
+      tree.ResetBit(kMustCleanup);
```

Comment:
If I recall correctly, @pcanal indicated to me that this workaround might not be necessary anymore.
It would be great if you could check if that's really the case and then remove it from the `TBufferMerger` tests.

Reply:
If this is true I will open another PR to modify all the tests in TBufferMerger, thank you for the hint!

Reply:
@vepadulano can you take care of this (in another PR)? e.g. here, for sure the ResetBit(kMustCleanup) is not needed to prevent lifetime issues because the TTree goes out of scope before the file that contains it (and TTree deregisters itself from the file at destruction time).

Reply:
Yeah sure!